Initial release

smittytone opened the pull request.

smittytone commented:
Kind of stuck on the tests side of this because it's so async

betzrhodes commented:
Not sure if this will help with test ideas, but here are some async usb test examples: https://github.com/electricimp/Usb/tree/master/tests/manual

smittytone commented:
@betzrhodes this is ready to go, I think. Written some tests, and they pass. See what you think.

betzrhodes commented on README.md:
```diff
@@ -0,0 +1,107 @@
+# USB Hub Driver 1.0.0 #
+
+This library provides adds basic support for USB hubs to Electric Imp’s existing [USB driver framework](https://github.com/electricimp/Usb).
```
typo - provides adds?

betzrhodes commented on README.md:
```diff
@@ -0,0 +1,107 @@
+
+This library provides adds basic support for USB hubs to Electric Imp’s existing [USB driver framework](https://github.com/electricimp/Usb).
+
+**Important** Though this library supports hubs, it does not yet support the hot-plugging of devices to that hub. Devices **must** be connected to the hub before the hub is connected to the imp. You can use the method [*checkports()*](#checkports) to verify the status of a hub's ports at any time.
```
This is a big limitation. It makes me wonder if we should not publish this as a library, but publish it along with our other example drivers. in the USB library repo.

betzrhodes replied:
If we do publish as a library then this should be a beta release (version less than 1.0.0)

betzrhodes commented on README.md:
````diff
@@ -0,0 +1,107 @@
+```squirrel
+agent.on("list.devices", function(ignored) {
+    if (hubDriver != null) {
+        local ports = hubDriver.checkPorts();
````
This feels more like a full example rather than a simple usage example for a method. Can we use this as a full example,  and just use this bit for the function usage example.

smittytone replied:
Moved it out of the Read Me to the /examples folder

betzrhodes commented on USB.hub.device.lib.nut:
```diff
@@ -0,0 +1,360 @@
+                               blob(0));
+
+        if (initDevices) {
+            // Wait 50ms
```
comment and sleep time mis-match. This would be better kept in a constant.

betzrhodes commented on tests/HubDriverTestCase1.test.nut:
```diff
@@ -0,0 +1,123 @@
+// ---------------------------------------------------------------------
+
+// Local includes
+@include __PATH__+"/../Documents/GitHub/Usb/USB.device.lib.nut"
```
these should be includes from github not a local file directory

betzrhodes commented on tests/HubDriverTestCase2.test.nut:
```diff
@@ -0,0 +1,149 @@
+// ---------------------------------------------------------------------
+
+// Local includes
+@include __PATH__+"/../Documents/GitHub/Usb/USB.device.lib.nut"
```
update to includes from github repo

betzrhodes commented on tests/README.md:
```diff
@@ -0,0 +1,42 @@
+
+## Setup ##
+
+In both cases, you will need to to connect your UB hub and attached peripherals(s) before running the tests.
```
typo? UB

betzrhodes commented on tests/README.md:
```diff
@@ -0,0 +1,42 @@
+
+In both cases, you will need to to connect your UB hub and attached peripherals(s) before running the tests.
+
+The tests require Electric Imp’s [impt](https://github.com/electricimp/imp-central-impt) tool and load in USH Host, FTDI and USB Hub drivers locally from the host machine. You should clone the [USB Host](https://github.com/electricimp/Usb) as well as the USB Hub repos and update the tests lines 44 through 46 to point to the named files on your system.
```
typo?  USH Host

no need to clone and update code - just include file from github source

smittytone commented:
OK. I've create a separate example, updated the @includes to GitHub locations and fixed the various typos. I've also enum'd many of the numeric literal appearing in the library - at least where it makes sense to do so.
Let me know whether you want to put this out here as a beta (I'll change the version number to 0.1.0) or shut thus repo down and move the code etc to the USB library repo.

betzrhodes commented:
Looks good. Until this gets more usage, let's move this to the USB library repo with the other drivers.

smittytone commented:
Moved the code over to https://github.com/electricimp/Usb/pull/18
Will close the dedicated repo down when I've completed testing.
